File: Northern Frontier Logistics/d1.py
import sys
def solve():
    data = list(map(int, sys.stdin.read().split()))
    idx, T = 0, data[idx]; idx += 1
    out = []
    for _ in range(T):
        N, D = data[idx], data[idx+1]; idx += 2
        W = data[idx:idx+N]; idx += N
        lo, hi = max(W), sum(W)
        while lo < hi:
            mid = (lo + hi) // 2
            cnt, cur = 1, 0
            for w in W:
                if cur + w > mid:
                    cnt += 1
                    cur = 0
                cur += w
            if cnt <= D:
                hi = mid
            else:
                lo = mid + 1
        out.append(str(lo))
    sys.stdout.write("\n".join(out))
if __name__ == "__main__":
    solve()

File: Northern Frontier Logistics/d1.cpp
#include <bits/stdc++.h>
using namespace std;
int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int T; cin >> T;
    while (T--) {
        int N, D; cin >> N >> D;
        vector<int> W(N);
        int lo = 0, hi = 0;
        for (int i = 0; i < N; ++i) {
            cin >> W[i];
            lo = max(lo, W[i]);
            hi += W[i];
        }
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            int cnt = 1, cur = 0;
            for (int w : W) {
                if (cur + w > mid) { ++cnt; cur = 0; }
                cur += w;
            }
            (cnt <= D) ? hi = mid : lo = mid + 1;
        }
        cout << lo << '\n';
    }
}